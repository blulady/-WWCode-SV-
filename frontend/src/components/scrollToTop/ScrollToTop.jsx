import React, { useState, useEffect, useCallback } from 'react';
import styles from './ScrollToTop.module.css';

const ScrollToTop = () => {
    const [isVisible, setIsVisible] = useState(false);
    const [isMobile, setIsMobile] = useState(false);
    const [scrollY, setScrollY] = useState(window.scrollY);

    const checkIsMobile = () => {
        const resolution = window.innerWidth;

        if (resolution <= 480) {
            setIsMobile(true);
        }
    };

    const handleScroll = useCallback(
        (e) => {
            const scrolled = document.documentElement.scrollTop;
            const window = e.currentTarget;

            // Comment out if statement and use comment below to see button 
            //if (scrollY > window.scrollY) {
            if (scrolled > 4340 && scrollY > window.scrollY) {
                setIsVisible(true);
            } else {
                setIsVisible(false);
            }

            setScrollY(window.scrollY);
        },
        [scrollY],
    );

    const scroll = () => {
        window.scrollTo({
            top: 0,
            behavior: 'smooth'
        });

        setIsVisible(false);
    };

    useEffect(() => {
        checkIsMobile();
    }, []);

    useEffect(() => {
        setScrollY(window.scrollY);

        window.addEventListener("scroll", handleScroll);

        return () => {
            window.removeEventListener("scroll", handleScroll);
        }
    }, [handleScroll]);

    return (
        <div>
            <button onClick={scroll}
            style={{display : isVisible && isMobile ? 'inline' : 'none'}}
            className={styles['scroll-to-top-button']} >
                Back to top <i className="fas fa-angle-double-up 2x"></i>
            </button>
        </div>
    );
}

export default ScrollToTop;
