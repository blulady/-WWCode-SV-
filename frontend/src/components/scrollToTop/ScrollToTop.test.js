import React from 'react';
import { render, fireEvent, act } from '@testing-library/react';
import ScrollToTop from './ScrollToTop';

describe('ScrollToTop', () => {
    test('it renders without crashing', () => {
        const { container } = render(<ScrollToTop />);
        expect(container).toMatchSnapshot();
    });

    test('it does not display button when scrollY > 4340 and user is scrolling up and not on mobile', () => {
        const { getByText } = render(<ScrollToTop />);
        fireEvent.scroll(global, { target: { scrollY: 4350 } });
        fireEvent.scroll(global, { target: { scrollY: 4341 } });
        const button = getByText(/Back to Top/i);

        expect(button).toHaveClass('scroll-to-top-not-visible');
    });

    test('it displays button when scrollY > 4340 and user is scrolling up and on mobile', () => {
        global.innerWidth = 414;
        global.dispatchEvent(new Event('resize'));

        const { container, getByText } = render(<ScrollToTop />);
        const scrollTopSpy = jest.spyOn(document.documentElement, 'scrollTop', 'get')
                                .mockImplementation(() => 4350);

        fireEvent.scroll(global, { target: { scrollY: 4350 } });
        fireEvent.scroll(global, { target: { scrollY: 4341 } });

        const button = getByText(/Back to Top/i);
        expect(button).toHaveClass('scroll-to-top-visible');
    });

    test('when button is pressed, it scrolls to top and does not display button', () => {
        global.innerWidth = 414;
        global.dispatchEvent(new Event('resize'));

        const { container, getByText } = render(<ScrollToTop />);
        const scrollTopSpy = jest.spyOn(document.documentElement, 'scrollTop', 'get')
                                 .mockImplementation(() => 4350);
        window.scrollTo = jest.fn();

        act(() => {
            fireEvent.scroll(global, { target: { scrollY: 4350 } });
            fireEvent.scroll(global, { target: { scrollY: 4341 } });
    
            const button = getByText(/Back to Top/i);
            fireEvent.click(button);    
        });

        setTimeout(() => {
            expect(global.scrollY).toBe(0);
            expect(button).toHaveClass('scroll-to-top-not-visible');
        }, 100);
    });
});