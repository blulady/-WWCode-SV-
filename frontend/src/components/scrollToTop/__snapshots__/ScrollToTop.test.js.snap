// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScrollToTop it renders without crashing 1`] = `
<div>
  <div>
    <button
      class="scroll-to-top-button scroll-to-top-not-visible"
    >
      Back to top 
      <i
        class="fas fa-angle-double-up 2x"
      />
    </button>
  </div>
</div>
`;
